### PR TITLE
[PSPE] [WIP] EntityMetadata

### PR DESCRIPTION
**Pull request by 7kasper**
Entitymetadata for PocketEdition
--------------------------
Currently working (metadata Server -> Client):
* Various tags (Sprinting, Sneaking, onFire, Invisible and Gliding)
* Air bubbles
* Nametags

_Priviously_
> * Implemented Client -> Server sneaking
> * Fixed error when sneaking (Updated listener with extra check for PE connection)
> * Messing with EntityMetadata~~
> 
> Well I tried to mess with it a bit, inside the preexisting remapper but couldn't get it to work on PE yet.
> However, the output (currently crudely given using System.out) seems to be correct.
> The smallest mistakes are often hardest to find.. Especially with packets
> 
> Help is welcome :wink:

### CODE REVIEW

**Comment by Shevchik** [comment]
The listener should check that version is pc, and that should probably be in the main branch.

**Comment by 7kasper** [comment]
@Shevchik I see you've implemented sneaking and changed the remapper. Nice!
I'll adapt the PE version, but there's probably still a mistake somewhere in how the data is send to PE on my part. To be continued..

**Comment by 7kasper** [comment]
Allright So the metadata system is half encoded in Little Endian and half in Big Endian. Some stuff with shorts have to be moved and the identifiers have to be in Big Endian.

Here's some code that works and sets the bubbles to 2/3:
``` Java
		//Testing proper format of meta in general..
		ClientBoundPacketData serializer = ClientBoundPacketData.create(PEPacketIDs.ENTITY_METADATA, version);
		VarNumberSerializer.writeSVarLong(serializer, entityId); //ID
		VarNumberSerializer.writeVarInt(serializer, 3); //Size
		VarNumberSerializer.writeVarInt(serializer, 0); //Flags
		VarNumberSerializer.writeVarInt(serializer, 7); //Type
		VarNumberSerializer.writeSVarLong(serializer, 0l); //Nothing.
		VarNumberSerializer.writeVarInt(serializer, 7); //Air
		VarNumberSerializer.writeVarInt(serializer, 1); //Type
		MiscSerializer.writeLShort(serializer, (short) 200);//2/3 the bubbles.
		VarNumberSerializer.writeVarInt(serializer, 38); //Leash Holder
		VarNumberSerializer.writeVarInt(serializer, 7); //Type
		VarNumberSerializer.writeVarLong(serializer, -1l); //Nothing.
		return RecyclableSingletonList.create(serializer);
}
```

**Comment by Shevchik** [comment]
All of that should be in a specific remapper, not in general one.

**Comment by 7kasper** [comment]
@Shevchik right now I've got sneaking, sprinting airbubbles and nametags working.
There is a bit of refractoring needed before going into the specific remapper.
What do you want me to change precisely?

File: src/protocolsupport/protocol/packet/middleimpl/clientbound/play/v_pe/PlayerInfo.java [hunk 1]
```diff
@@ -25,9 +25,9 @@
 				for (Info info : infos) {
 					MiscSerializer.writeUUID(serializer, info.uuid);
 					VarNumberSerializer.writeSVarInt(serializer, 0); //entity id
-					StringSerializer.writeString(serializer, version, info.getName());
+					StringSerializer.writeString(serializer, version, info.username);
 					StringSerializer.writeString(serializer, version, "Standard_Steve");
-					ByteArraySerializer.writeByteArray(serializer, version, PESkin.STEVE);
+					ByteArraySerializer.writeByteArray(serializer, version, PESkin.STEVE); //Meh.
```
Comment (NycuRO) [inline src/protocolsupport/protocol/packet/middleimpl/clientbound/play/v_pe/PlayerInfo.java]: Default skin is Steve and Alex...

Comment (7kasper) [inline src/protocolsupport/protocol/packet/middleimpl/clientbound/play/v_pe/PlayerInfo.java]: You're right, but we've only implemented Steve for now. All players are Steves :)

File: src/protocolsupport/ProtocolSupport.java [hunk 2]
```diff
@@ -65,7 +65,7 @@ public void onLoad() {
 			BlockStorageReader.init();
 			ServerPlatform.get().inject();
 			PESkin.init();
-			server = new MCPEServer(19132);
+			server = new MCPEServer(2222);
```
Comment (NycuRO) [inline src/protocolsupport/ProtocolSupport.java]: @7kasper Can you add costumizable?

Comment (7kasper) [inline src/protocolsupport/ProtocolSupport.java]: @NycuRO I could but that's up to @Shevchik. For now at least you'll have to change that line before you compile Protocol Support.

Comment (NycuRO) [inline src/protocolsupport/ProtocolSupport.java]: Ok.

**Comment by 7kasper** [comment]
So there's a bit of a problem with the way Meta is updated on the PC version. The pc version only sends stuff that's changed. That wouldn't be a problem, but many metafields in PC are contained in the first field for PE as tags, that need to be send on every metaupdate. To know wich tags to set you need access to the entire PC metadata, but that's only send as changes. So, I think we need to keep track of ALL meta fields in the NetworkDataCache.

**Comment by NycuRO** [comment]
@7kasper First please. Anyone use Spigot Plugins , not PSPE.

**Comment by Shevchik** [comment]
Meta remapper needs a rework, so merging this will have to wait.

**Comment by NycuRO** [comment]
@7kasper Can you use this for Metadata :D : https://github.com/Hydreon/Steadfast2/blob/1.1.0_beta/src/pocketmine/entity/Entity.php

**Comment by DaMatrix** [comment]
@NycuRO we're mostly using Nukkit for basing code from because Java

**Comment by 7kasper** [comment]
Tbh I scramble from everywere, but Nukkit is indeed particulary useful sometimes. (It's very simular to Pocketmine of course).

This PR will be updated with alot (probably most) of the entitymeta implementation for PE, but I will not work on it before the meta remapper is finished, because many features will depend on that.

**Comment by NycuRO** [comment]
Abandonated? @7kasper

**Comment by DaMatrix** [comment]
No, it's just not being worked on ATM. There's other things to do than MCPE servers.

**Pull request closed without merging**